tumblr/kanvas-ios
# Simplified permission flow

JIRA: [ICORE-4812](https://jira.tumblr.net/browse/ICORE-4812)
P2: https://wp.me/pbzQyC-248

## What it does

This changes how we ask the user for camera and microphone permissions. Apple recently rejected our review as they didn't want an intermediary screen where users can choose whether to display apples official permission requests.

**__Everything here has been reviewed__** for orangina as part of this [PR](https://github.tumblr.net/TumblrMobile/orangina/pull/21649/files) and is currently being used within Orangina by pointing at a copy of this branch.

| Before | After |
| --- | --- |
| ![Simulator Screen Recording - iPhone 8 - 2022-04-22 at 10 11 38](https://user-images.githubusercontent.com/18557720/164627064-bbe635b8-e2f1-415c-92f2-88187c5f5671.gif) | ![Simulator Screen Recording - iPhone 8 - 2022-04-22 at 10 13 26](https://user-images.githubusercontent.com/18557720/164627116-6bb0e2da-26ca-4784-b471-797abf6acdcc.gif)| 

## How to test
This might not need testing as it's already been tested when added to Orangina, although in case the example app behaves significantly differently and you'd like to check I've included the testing instructions from the original PR. There's also another PR that will be merged in after this one adding automated tests and removing code, it might be better to test [this PR](https://github.com/tumblr/kanvas-ios/pull/134) instead

Use a real device, resetting the app permission settings on simulator is a bit tricky.

#### Unknown permissions
* Goto `Settings > General > Reset > Reset Location & Privacy`
* Open photo screen and two alerts should appear

#### Accepted permissions
* Goto `Settings > General > Reset > Reset Location & Privacy`
* Open photo screen and accept both permission alerts
* Result: Camera and mic should be fully accessible
* Go back and open screen
Result: Camera and mic should be fully accessible

#### Denied permissions - 
* Goto `Settings > General > Reset > Reset Location & Privacy`
* Open photo screen
* Deny Camera and Mic access
* Result: Settings redirect screen should appear and no access to camera is granted
* Go back and reopen screen
* Result: No permission alerts should appear. Settings redirect screen should appear and no access to camera is granted
* Kill app and open screen
* Result: Settings redirect screen should appear and no access to camera is granted
* repeat these steps this time only selecting one of the camera or mic permissions to reject
